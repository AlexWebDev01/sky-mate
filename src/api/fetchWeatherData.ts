import { Coordinates } from "../context/GlobalContext.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: string
) => {
  console.log("COORDINATES:", coordinates);

  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data = await response.json();

  return data;
};
