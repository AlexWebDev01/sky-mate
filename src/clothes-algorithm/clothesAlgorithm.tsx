import { TEMPERATURE_RANGES, CLOTHES_ADVICES } from "./clothesAlgorithmData";

const getTempRange = (temp: number): string => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  return "";
};

export const calculateClothesAdvice = (
  weatherCondition: string,
  temp: number
) => {
  const clothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return { adviceData: clothesAdvices[tempRangeName], tempRangeName };
  } else {
    console.log(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return null;
  }
};
