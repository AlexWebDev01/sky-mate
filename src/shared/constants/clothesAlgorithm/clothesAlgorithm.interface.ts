export enum WeatherConditions {
  rain = 'rain',
  snow = 'snow',
  sun = 'sun',
  clear = 'clear',
  clouds = 'clouds',
  fog = 'fog',
  dust = 'dust',
  mist = 'mist',
}

export enum TemperatureRangeNames {
  DANGEROUSLY_FREEZING = 'DANGEROUSLY_FREEZING',
  EXTREMELY_FREEZING = 'EXTREMELY_FREEZING',
  FREEZING = 'FREEZING',
  VERY_COLD = 'VERY_COLD',
  COLD = 'COLD',
  SLIGHTLY_COLD = 'SLIGHTLY_COLD',
  SLIGHTLY_FRESH = 'SLIGHTLY_FRESH',
  FRESH = 'FRESH',
  WARM = 'WARM',
  VERY_WARM = 'VERY_WARM',
  HOT = 'HOT',
  VERY_HOT = 'VERY_HOT',
  DANGEROUSLY_HOT = 'DANGEROUSLY_HOT',
}

export interface TempRanges {
  name: TemperatureRangeNames;
  lowestTemp: number;
  highestTemp: number;
}

export interface ClothesAdvice {
  clothesStyle: ClothesStyle;
  clothesList: string[];
  clothesDescription: string;
}

export interface ClothesAdviceWithTemp extends ClothesAdvice {
  tempRangeName: TemperatureRangeNames | null;
}

export type ClothesAdvices = {
  [K in TemperatureRangeNames]?: ClothesAdvice;
};

export interface ClothesAdvicesMap {
  rain: ClothesAdvices;
  sun: ClothesAdvices;
  clear: ClothesAdvices;
  clouds: ClothesAdvices;
  snow: ClothesAdvices;
  fog: ClothesAdvices;
  dust: ClothesAdvices;
  mist: ClothesAdvices;
}

export enum ClothesStyle {
  basic = 'Basic look',
  protect = 'Protect look',
  elegant = 'Elegant look',
  home = 'Home look',
  noLook = 'Oops, no look!',
}
