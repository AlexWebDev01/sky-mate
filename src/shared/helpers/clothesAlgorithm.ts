import {
  ClothesAdviceWithTemp,
  ClothesAdvices,
  TEMPERATURE_RANGE_NAMES,
  WeatherConditions,
  clothesStyle,
} from "../constants/clothesAlgorithm/clothesAlgorithm.interface";
import {
  TEMPERATURE_RANGES,
  CLOTHES_ADVICES,
} from "../constants/clothesAlgorithm/clothesAlgorithmData";

const getTempRange = (temp: number): TEMPERATURE_RANGE_NAMES => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  throw new Error(`There is no temperature range for provided temp: ${temp}`);
};

export const calculateClothesAdvice = (
  weatherCondition: WeatherConditions,
  temp: number
): ClothesAdviceWithTemp => {
  console.log(weatherCondition);
  const clothesAdvices: ClothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return {
      ...clothesAdvices[tempRangeName],
      tempRangeName,
    } as ClothesAdviceWithTemp;
  } else {
    console.error(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return {
      clothesStyle: clothesStyle.noLook,
      clothesList: [""],
      clothesDescription:
        "Ahh, the weather is so unpredictable, we don't have any advice for you! Try following your intuition or simply look at the sky.",
      tempRangeName: null,
    };
  }
};
