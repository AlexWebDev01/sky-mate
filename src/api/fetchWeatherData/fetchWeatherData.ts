import { Units, Coordinates, WeatherData } from "./fetchWeatherData.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: Units
): Promise<WeatherData> => {
  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data: WeatherData = await response.json();

  return data;
};
