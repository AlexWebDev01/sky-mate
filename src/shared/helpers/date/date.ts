import { Coordinates } from "../../../api/fetchWeatherData/fetchWeatherData.interface";
import { FullWeekday, ShortWeekday } from "./date.interface";

export function epochToLocalTime(
  epochTime: number,
  offsetTime: number
): string {
  // create a new Date object using the epoch time multiplied by 1000 (to convert from seconds to milliseconds)
  const date = new Date(epochTime * 1000);

  // calculate the local time offset in milliseconds
  const localTimeOffset = offsetTime * 1000;

  // create a new Date object that represents the local time
  const localDate = new Date(date.getTime() + localTimeOffset);

  // return the local time as a string in the format HH:MM
  return localDate.toISOString().substring(11, 16);
}

// Format: dd.mm
export function calculateLocalDate(unixTimestamp: number): string {
  const fullDate = new Date(unixTimestamp * 1000);
  const month = fullDate.getMonth().toString();
  const date = fullDate.getDate().toString();

  return `${date.padStart(2, "0")}.${month.padStart(2, "0")}`;
}

// Format: dd month
export function calculateLocalDateAnotherFormat(unixTimestamp: number): string {
  const monthNames = [
    "jan",
    "feb",
    "mar",
    "apr",
    "may",
    "jun",
    "jul",
    "aug",
    "sept",
    "oct",
    "nov",
    "dec",
  ];
  const fullDate = new Date(unixTimestamp * 1000);
  const month = fullDate.getMonth();
  const monthFormatted = monthNames[month];
  const date = fullDate.getDate().toString();

  return `${date} ${monthFormatted}`;
}

export function calculateLocalFormattedDay(
  unixTimestamp: number
): ShortWeekday | "" {
  const date = new Date(unixTimestamp * 1000);
  const dayOfWeek = date.getDay();
  switch (dayOfWeek) {
    case 0:
      return ShortWeekday.Sun;
    case 1:
      return ShortWeekday.Mon;
    case 2:
      return ShortWeekday.Tue;
    case 3:
      return ShortWeekday.Wed;
    case 4:
      return ShortWeekday.Thu;
    case 5:
      return ShortWeekday.Fri;
    case 6:
      return ShortWeekday.Sat;
    default:
      return "";
  }
}

export function calculateLocalDay(unixTimestamp: number): FullWeekday | "" {
  const date = new Date(unixTimestamp * 1000);
  const dayOfWeek = date.getDay();
  switch (dayOfWeek) {
    case 0:
      return FullWeekday.Sunday;
    case 1:
      return FullWeekday.Monday;
    case 2:
      return FullWeekday.Tuesday;
    case 3:
      return FullWeekday.Wednesday;
    case 4:
      return FullWeekday.Thursday;
    case 5:
      return FullWeekday.Friday;
    case 6:
      return FullWeekday.Saturday;
    default:
      return "";
  }
}

export function separateCoordinates(coordinatesString: string): Coordinates {
  const [latitude, longitude] = coordinatesString.split(",");
  return {
    lat: Number(latitude.trim()),
    lon: Number(longitude.trim()),
  };
}
