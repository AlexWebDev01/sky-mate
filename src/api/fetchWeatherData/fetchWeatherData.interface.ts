import { WeatherConditions } from "../../shared/constants/clothesAlgorithm/clothesAlgorithm.interface";

export interface WeatherData {
  current: CurrentWeatherData;
  daily: DailyWeatherData[];
  hourly: HourlyWeatherData[];
  lat: number;
  lon: number;
  timezone: string;
  timezone_offset: number;
}

export interface CurrentWeatherData extends BaseWeatherData {
  sunrise: number;
  sunset: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface HourlyWeatherData extends BaseWeatherData {
  pop: number;
  wind_gust: number;
  visibility: number;
  temp: number;
  feels_like: number;
}

export interface DailyWeatherData extends BaseWeatherData {
  feels_like: { day: number; night: number; eve: number; morn: number };
  moon_phase: number;
  moonrise: number;
  moonset: number;
  pop: number;
  rain: number;
  summary: string;
  sunrise: number;
  sunset: number;
  temp: {
    day: number;
    eve: number;
    max: number;
    min: number;
    morn: number;
    night: number;
  };
  wind_gust: number;
}

interface BaseWeatherData {
  clouds: number;
  dew_point: number;
  dt: number;
  humidity: number;
  pressure: number;
  uvi: number;
  weather: Weather[];
  wind_deg: number;
  wind_speed: number;
}

export interface Weather {
  description: string;
  icon: string;
  id: number;
  main: WeatherConditions;
}

export interface Coordinates {
  lat: number;
  lon: number;
}

export enum Units {
  standard = "standard",
  metric = "metric",
  imperial = "imperial",
}
