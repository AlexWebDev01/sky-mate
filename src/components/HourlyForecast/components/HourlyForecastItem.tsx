import { HourlyWeatherData } from '../../../api/fetchWeatherData/fetchWeatherData.interface';
import { epochToLocalTime } from '../../../shared/helpers/date/date';

interface Props {
  item: HourlyWeatherData;
  timezoneOffset: number;
}

const HourlyForecastItem = ({ item, timezoneOffset }: Props) => {
  return (
    <div className='day'>
      <div>{Math.round(item.temp)}&deg;</div>
      <div className='line'></div>
      <div>{epochToLocalTime(item.dt, timezoneOffset).slice(0, 2)}</div>
    </div>
  );
};

export default HourlyForecastItem;
