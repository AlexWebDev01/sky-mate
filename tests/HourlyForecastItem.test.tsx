import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import HourlyForecastItem from '@components/HourlyForecast/components/HourlyForecastItem';
import { HourlyWeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';
import { epochToLocalTime } from '@shared/helpers/date/date';
import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';

jest.mock('@shared/helpers/date/date', () => ({
  epochToLocalTime: jest.fn(),
}));

const mockEpochToLocalTime = epochToLocalTime as jest.MockedFunction<
  typeof epochToLocalTime
>;

jest.mock('@shared/assets/icons/link-arrow.svg', () => {
  return () => <svg data-testid='mocked-arrow'></svg>;
});

describe('HourlyForecastItem Component', () => {
  const mockItem: HourlyWeatherData = {
    clouds: 75,
    dew_point: 10,
    dt: 1625235600,
    humidity: 65,
    pressure: 1013,
    uvi: 0.5,
    weather: [
      {
        id: 500,
        main: WeatherConditions.rain,
        description: 'light rain',
        icon: '10d',
      },
    ],
    wind_deg: 270,
    wind_speed: 5.1,
    pop: 0.2,
    wind_gust: 7.0,
    visibility: 10000,
    temp: 20.5,
    feels_like: 20,
  };

  const mockTimezoneOffset = -14400; // UTC-4

  beforeEach(() => {
    mockEpochToLocalTime.mockReturnValue('14:00');
  });

  test('renders temperature correctly', () => {
    render(
      <HourlyForecastItem item={mockItem} timezoneOffset={mockTimezoneOffset} />
    );
    const tempElement = screen.getByText('21Â°');
    expect(tempElement).toBeInTheDocument();
  });

  test('renders the time correctly', () => {
    render(
      <HourlyForecastItem item={mockItem} timezoneOffset={mockTimezoneOffset} />
    );
    const timeElement = screen.getByText('14');
    expect(timeElement).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <HourlyForecastItem item={mockItem} timezoneOffset={mockTimezoneOffset} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
