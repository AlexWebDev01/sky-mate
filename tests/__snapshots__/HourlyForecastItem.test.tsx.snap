// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourlyForecastItem Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="day"
  >
    <div>
      21Â°
    </div>
    <div
      class="line"
    />
    <div>
      14
    </div>
  </div>
</DocumentFragment>
`;
